File: solutions/advanced/DP on a Tree/SumOfDistancesInTree.cpp
/*
    Sum of Distances in Tree

    https://leetcode.com/problems/sum-of-distances-in-tree/description/
*/

#include <iostream>
#include <vector>
using namespace std;
const int N = 100000+21;
vector<int> graph[N];
int dp[N][2];
int roz[N];
int n, a, b;
int answer = 0;
void dfs1(int v, int prev)
{
    for (int i=0; i<graph[v].size(); i++)
    {
        int u = graph[v][i];
        if (u != prev)
        {
            dfs1(u, v);
            roz[v] += roz[u];
            dp[v][0] += dp[u][0] + roz[u];
        }
    }
    roz[v]++;

}

void dfs2(int v, int prev, int prev_val)
{
    dp[v][1] = dp[v][0] + prev_val + (n - roz[v]);

    for (int i=0; i<graph[v].size(); i++)
    {
        int u = graph[v][i];
        if (u != prev)
        {
            dfs2(u, v, dp[v][1] - dp[u][0] - roz[u]);
        }
    }
    roz[v]++;

}
int main()
{
    cin>>n;
    for (int i=1; i<n; i++)
    {
        cin>>a>>b;
        graph[a].push_back(b);
        graph[b].push_back(a);
    }

    dfs1(0, -1);
    dfs2(0, -1, 0);
    for (int i=0; i<n; i++)
	    cout<<dp[i][1]<<endl;
}

File: solutions/advanced/DP on a Tree/SumOfDistancesInTree.py

n = int(input())
graph = [[] for _ in range(n)]
dp = [[0, 0] for _ in range(n)]  # dp[v][0] = sum of distances in subtree, dp[v][1] = final answer for node v
roz = [0] * n  # roz[v] = number of nodes in subtree rooted at v

# Read edges
for _ in range(n - 1):
    a, b = map(int, input().split())
    graph[a].append(b)
    graph[b].append(a)

def dfs1(v, parent):
    for u in graph[v]:
        if u != parent:
            dfs1(u, v)
            roz[v] += roz[u]
            dp[v][0] += dp[u][0] + roz[u]
    roz[v] += 1

def dfs2(v, parent, prev_val):
    print(v, parent, prev_val);
    dp[v][1] = dp[v][0] + prev_val + (n - roz[v])
    for u in graph[v]:
        if u != parent:
            # Recalculate the value passed to child based on the current node's value
            dfs2(u, v, dp[v][1] - (dp[u][0] + roz[u]))

dfs1(0, -1)
dfs2(0, -1, 0)

for i in range(n):
    print(dp[i][1])
